DOCS-6510 Changes (#16)
mulesoft/docs-functional-monitoring: modules/ROOT/pages/bat-bdd-reference.adoc

A test suite must contain at least one test (.dwl) file.

== Keeping Confidential Information Secure in Tests

> In your tests, you can access shared secrets in Anypoint Secrets Manager, which allows you to store securely sensitive information, such as a password, authentication token, endpoint URL, webhook URL, or something else. You create aliases for your shared secrets with the `bat grant` command. The BAT CLI stores aliases in your `bat.yaml` files in your test suites. You place aliases in your tests and, at runtime, the BAT CLI looks up the shared secrets by means of the aliases.

Reviewer: I suggested a rewording to the beginning of the first sentence below in a previous edit; I'll leave this one as-is and let you decide about where "in your tests" belongs....
```suggestion
In your tests, you can access shared secrets in Anypoint Secrets Manager, which enables you to securely store sensitive information, such as a password, authentication token, endpoint URL, or webhook URL. 

You create aliases for your shared secrets with the `bat grant` command. The BAT CLI stores aliases in your `bat.yaml` files in your test suites. You place aliases in your tests and, at runtime, the BAT CLI looks up the shared secrets by means of the aliases.
```

[IMPORTANT]
====
In a given test suite, you can use aliases for shared secrets only if you plan to run the tests in that suite from a private location.
====

> 

Reviewer: I am sure you have thought of this but in performing due diligence, I have to ask: is there any way to reduce the following steps to just one user action apiece, with less narrative underpinning?

Reviewer: Also, I believe that the coding for substeps in AsciiDoc is ". ." rather than the letters you have used here.

Reviewer: GLOBALLY break multisentence steps into a new para after the first sentence.

Author: 1. I don't see that extra info as narration. I see it as necessary because of the complexity of the task.
2. No, I'm specifying in Asciidoc that I want to use the lowercase letters at the second level, not integers again.
3. Resolved here.

> . Store the information () as a shared secret in Anypoint Secrets Manager.

Reviewer: I don't understand why " ( ) " is included here.

Author: I was typing too exuberantly.

> . In Anypoint Secrets Manager, create a shared secret as a symmetric key. If the info that you want to store as a shared secret is currently in plain text, then you must encode it in Base64 when you create the secret.

Reviewer: ```suggestion
. In Anypoint Secrets Manager, create a shared secret as a symmetric key. 
If the information that you want to store as a shared secret is currently in plain text, then you must encode it in Base64 when you create the secret.
```

> a. Open a secrets group or create a new one. The group must be created in the same environment that you are using in the BAT CLI. To find out which environment you are currently using in the BAT CLI, run the command `bat whoami`. In the output, there is the ID for the environment. Run the command `bat environment ls` to list the environments that you have access to. Match the ID from the `bat whoami` command with one of the environments listed. If you need to switch to the environment that your secrets group is in, run the command `bat environment switch name`, where `name` is the name of the environment.

Reviewer: This step really should end after the second sentence. "Create  New Secrets Group" should be a separate topic.
```suggestion
.. Open a secrets group or create a new one. 
The group must be created in the same environment that you are using in the BAT CLI. 

To find out which environment you are currently using in the BAT CLI, run the command `bat whoami`. The output contains the ID for the environment. Run the command `bat environment ls` to list the environments that you have access to. Match the ID from the `bat whoami` command with one of the environments listed. 

If you need to switch to the environment that your secrets group is in, run the command `bat environment switch name`, where `name` is the name of the environment.
```

Author: Sorry, but I disagree. That sounds like a topic for the Secrets Manager documentation. Users don't create them in AFM.

> b. Select Shared Secret on the left side of the screen.

Reviewer: ```suggestion
.. Select *Shared Secret* on the left side of the screen.
```

Author: Resolved, except for the change to the numbering.

> c. In the *Type* field, select *Symmetric Key*.

Reviewer: This is a field that offers choices rather than taking typed or pasted input?

Author: Yes.

d. In the *Key* field, paste your license key encoded as a Base64 string.
e. Paste the same Base64 string into the *Confirm Key* field.
. Copy the name of the new shared secret.

> . At a command prompt, run the `bat grant` command, specifying an alias for the shared secret. When you run this command, the BAT CLI creates a section named `secrets` in your test suite's `bat.yaml` file, if the section does not already exist. In that section, the BAT CLI adds these indented lines:

Reviewer: ```suggestion
. At a command prompt, run the `bat grant` command, specifying an alias for the shared secret. 
When you run this command, the BAT CLI creates a section named `secrets` in your test suite's `bat.yaml` file, if the section does not already exist. In that section, the BAT CLI adds these indented lines:
```

Reviewer: Do we really have to document what the CLI does in the background?
AND the syntax values involved?
As with Step 1a., I think this information probably belongs somewhere outside of the task steps....

Author: Yes, we do need to document that. These topics are for the manual process of creating test suites. Users manually create the bat.yaml file for their test suite, so they need to know why content is added to it, so that they are not surprised by it.

+
----
  alias:
    secretId: "secret-ID"
----
+

> alias:: This is the alias that you specified in the `bat grant` command.

Reviewer: ```suggestion
alias:: The alias that you specified in the `bat grant` command.
```

Author: Resolved.

> secret-ID:: This is the ID of the secret within Anypoint Secrets Manager. This ID does not appear in ASM, so there is no way for someone looking in your `bat.yaml` file to associate the ID with any particular secret. The BAT CLI uses this ID to look up the secret that you associated with the alias.

Reviewer: ```suggestion
secret-ID:: The ID of the secret within Anypoint Secrets Manager. This ID does not appear in Anypoint Secrets Manager, so there is no way for someone looking in your `bat.yaml` file to associate the ID with any particular secret. The BAT CLI uses this ID to look up the secret that you associated with the alias.
```

Author: Resolved.

. Use the alias in a test by storing it in a variable, as in this example:
+
----
import * from bat::BDD
import * from bat::Assertions

var mySecret = String = secret(secretAlias='myAlias') default 'notFound'

---
suite(parts="Simple Suite") in [
  it must 'answer 200' in [
    GET '$(config.url)?token=$(mySecret)' with {} assert [
      $.response.status mustEqual 200
    ] execute
  ]
]
----

> +

Reviewer: Do we really need to define the syntax of the example here?

Author: Yes, I think so.

> `mySecret`:: The variable in which the secret is stored in this example. The variable is declared as a String.

Reviewer: ```suggestion
`mySecret`:: The variable in which the secret is stored in this example. The variable is declared as a string.
```

Author: "String" with a capital S is shorthand for "String object".

`secret(secretAlias='myAlias')`:: The `secret()` function requires the input `secretAlias='name-of-alias')`, where `name-of-alias` is the alias of the shared secret that you want to use.
`default 'notFound'`:: The message to show if a shared secret with the specified alias does not exist in Anypoint Secrets Manager.

> `$(mySecret)`:: The use of the variable `mySecret` to supply the authentication token for the endpoint at the URL that is defined in the config file that was specified when the `bat` command to run the test suite was entered. Suppose that the config file was this:

Reviewer: ```suggestion
`$(mySecret)`:: The use of the variable `mySecret` to supply the authentication token for the endpoint at the URL that is defined in the configuration file that was specified when the `bat` command to run the test suite was entered. 

For example, suppose that the configuration file was this:
```

+
.The example config file dev-environment.dwl
----
{
  url: "https://dev.yourcompany.com/shoppingCart"
}
----
+

> Then, the command to run this test suite, when the command is run in the directory in which the `bat.yaml` file for the test suite is located and the default profile is used, is this:

Reviewer: ```suggestion
In this case, the command to run this test suite, when the command is run in the directory in which the `bat.yaml` file for the test suite is located and the default profile is used, is this:
```
Is that long parenthetical in the middle of this sentence really necessary to audience understanding?

A test suite must contain at least one test (.dwl) file.

== Keeping Confidential Information Secure in Tests

> In your tests, you can access shared secrets in Anypoint Secrets Manager, which allows you to store securely sensitive information, such as a password, authentication token, endpoint URL, webhook URL, or something else. You create aliases for your shared secrets with the `bat grant` command. The BAT CLI stores aliases in your `bat.yaml` files in your test suites. You place aliases in your tests and, at runtime, the BAT CLI looks up the shared secrets by means of the aliases.

Reviewer 2: if the `bat grant` command was previously executed, we will reuse the previous generated grant. It can be forced to re-generate the grant using the `--forceUpdate ` option of the command.

Author: Resolved.

[IMPORTANT]
====
In a given test suite, you can use aliases for shared secrets only if the following two conditions are true:

* You plan to run the tests in that suite from a private location.

> * The APIs that your test suite tests are deployed in CloudHub through Anypoint Virtual Private Cloud.

Reviewer 2: This is not correct, you can use secrets to test any API (deployed anywhere), as long as the Location is private.

Author: Fixed.

====

. Store the information () as a shared secret in Anypoint Secrets Manager.

> . In Anypoint Secrets Manager, create a shared secret as a symmetric key that uses Base64 encoding.

Reviewer 2: If the secret you want to save is not Symetric, then you can encode it with base 64, and when we resolve it decoded.

Author: Resolved.

a. Open a secrets group or create a new one.
b. Select Shared Secret on the left side of the screen.
c. In the *Type* field, select *Symmetric Key*.
d. In the *Key* field, paste your license key encoded as a Base64 string.
e. Paste the same Base64 string into the *Confirm Key* field.
. Copy the name of the new shared secret.

> . At a command prompt, run the `bat grant` command, specifying an alias for the shared secret. When you run this command, the BAT CLI creates a section named `secrets` in your test suite's `bat.yaml` file, if the section does not already exist. In that section, the BAT CLI adds these indented lines:

Reviewer 2: It's important to have configured in the CLI the same environment as the Secret Group (in CSM) before execution the grant command.
To See which environment you can use the `bat whoami` command. And `bat environment ls` `bat environment switch` to list and switch between environments.

Author: Resolved.

  it must 'answer 200' in [
    GET '$(config.url)?token=$(mySecret)' with {} assert [
      $.response.status mustEqual 200
    ] execute
  ]
]
----
+
`mySecret`:: The variable in which the secret is stored in this example. The variable is declared as a String.
`secret(secretAlias='myAlias')`:: The `secret()` function requires the input `secretAlias='name-of-alias')`, where `name-of-alias` is the alias of the shared secret that you want to use.
`default 'notFound'`:: The message to show if a shared secret with the specified alias does not exist in Anypoint Secrets Manager.
`$(mySecret)`:: The use of the variable `mySecret` to supply the authentication token for the endpoint at the URL that is defined in the config file that was specified when the `bat` command to run the test suite was entered. Suppose that the config file was this:
+
.The example config file dev-environment.dwl
----
{
  url: "https://dev.yourcompany.com/shoppingCart"
}
----
+
Then, the command to run this test suite, when the command is run in the directory in which the `bat.yaml` file for the test suite is located and the default profile is used, is this:
+
----
bat --config=dev-environment.dwl
----

> At runtime, the BAT CLI replaces the variable with the shared secret that it finds associated with the specified alias in Anypoint Secrets Manager.

Reviewer 2: We're missing the option to use secrets for the reporters:
```
suite:
  name: "Hello World Suite"
files:
  - file: "tests/HelloWorld.dwl"
    softFail: false
reporters:
  - type: "HTML"
    outFile: "/tmp/HTML.html"

  - type: "Email"
    options:
      EMAILS: <email1>

secrets:
  mySecret:
    secretId: "cc4375e1-da56-443c-bf4f-7faaa9f11cd2"
  mySecret2:
    secretId: "cc4375e1-da56-gg3c-bf4f-7faaa9f10cd2"
  email1:
    secretId: "c2b881e0-c5fe-4050-813b-7ebb386b0592"

```

Author: No, that is in a different topic.
